### ISSUE TITLE
Problems with predictions on MacBook Air with M1 chip in Java project based on Maven (even on onnxruntime 1.11.0 - latest at this moment) - no osx-aarch64 dir in the native runtimes dir

### ISSUE THREAD

**Issue by pwittchen** [microsoft/onnxruntime#11054]
Hi,

I reported a bug in #10874 regarding the fact that onnxruntime does not work on MacBook Air with M1 chip in Java project based on Maven. There is also information that this bug is fixed in PR #10981 I upgraded dependency to onnxruntime to version `1.11.0` and **I still have exactly the same bug as I reported in that issue** regardless of that upgrade.

Calling a function:

```java
final OrtEnvironment environment = OrtEnvironment.getEnvironment();
```

causes an error:

```
java.lang.NoClassDefFoundError: Could not initialize class ai.onnxruntime.OrtEnvironment
```

I see that in project dependencies there is: 

`/onnxruntime/1.11.0/onnxruntime-1.11.0.jar!/ai/onnxruntime/native/osx-x64`

but error occurs anyway

Moreover, when I execute a single test instead of my whole suite, then I get an error:

```
java.lang.UnsatisfiedLinkError: no onnxruntime in java.library.path
```

Probably, the source of the bug is here:

https://github.com/microsoft/onnxruntime/blob/74eaaad768e0a5ad5e65b818605383c73dece509/java/src/main/java/ai/onnxruntime/OnnxRuntime.java#L103

in the method `initOsArch()`. I debugged it and in my case I see the following params:
- detectedOs = `osx`
- arch = `aarch64`

final result of os arch: `osx-aarch64`

and then in `init()` method:

```java
libraryDirPathProperty = System.getProperty(ONNXRUNTIME_NATIVE_PATH)
```

`libraryDirPathProperty` is `null` where `ONNXRUNTIME_NATIVE_PATH` is `onnxruntime.native.path`

Next, in ` extractFromResources(String library)` method, `resourcePath` is:

```
/ai/onnxruntime/native/osx-aarch64/libonnxruntime_providers_shared.dylib
```

and there is no such path.

In the path `ai/onnxruntime/native` there are only the following dirs:
- `linux-x64`
- `osx-x64`
- `win-x64`

but there is no `osx-aarch64`, which is detected as my arch on MacBook Air with M1 chip

**Comment by Craigacp** [comment]
I usually build it with `build.sh --update --build --build_java --config Release`. Did you recently upgrade XCode? That can cause weird issues when you need to re-accept various licenses and is usually the source of various build issues for me on macOS.

**Comment by Craigacp** [comment]
ONNX Runtime uses the compilers & system headers it installs, but if it's working for iOS development it should have enabled everything. Try completely deleting the `build` directory and rerunning the build.sh command.

**Comment by pwittchen** [comment]
I used your params from the previous comment and it's building :-), so I guess this is the correct way to do it. I'll post an update when the build will finish.

**Comment by pwittchen** [comment]
Ok, it was built on my machine.

**Comment by Craigacp** [comment]
I think it'll land in the next release whenever that comes, but `1.11.0` will not be updated.

**Comment by pwittchen** [comment]
I meant a new version - not updating the current one, but ok, I'll wait for the release.

**Comment by Craigacp** [comment]
> I meant a new version - not updating the current one, but ok, I'll wait for the release.

I believe ORT is on a quarterly release schedule for feature releases.

@faxu any chance of Java macOS/arm64 binaries (#10981) landing in a `1.11` point release?

**Comment by faxu** [comment]
> > I meant a new version - not updating the current one, but ok, I'll wait for the release.
> 
> I believe ORT is on a quarterly release schedule for feature releases.
> 
> @faxu any chance of Java macOS/arm64 binaries (#10981) landing in a `1.11` point release?

Targeting 1.12 for including this build in the Maven pkg.

**Comment by pwittchen** [comment]
Thanks for the reply @faxu. Can you tell when can I expect 1.12 release?

**Comment by faxu** [comment]
> Thanks for the reply @faxu. Can you tell when can I expect `1.12` release?

We just released 1.11, so it will be a few months.

**Comment by pwittchen** [comment]
Ok, thank you for the reply. I'll wait for that. Luckily, running my project with onnxruntime on Mac is not critical issue at the moment.

**Comment by Craigacp** [comment]
I don't believe there are nightly builds for Java (or ORT in general), but you should be able to build it yourself on a Linux aarch64 machine. It'll build a jar with just support for that platform, but you can copy in the other platform libraries from the version on Maven Central to make a jar for internal use if you want (make sure you build Linux aarch64 from the right version tag if you want to do this). I don't frequently build it on Linux aarch64, but I build everything on macOS aarch64 last night while working on a feature.

**Comment by lanking520** [comment]
> I don't believe there are nightly builds for Java (or ORT in general), but you should be able to build it yourself on a Linux aarch64 machine. It'll build a jar with just support for that platform, but you can copy in the other platform libraries from the version on Maven Central to make a jar for internal use if you want (make sure you build Linux aarch64 from the right version tag if you want to do this). I don't frequently build it on Linux aarch64, but I build everything on macOS aarch64 last night while working on a feature.

Thanks for your response. Is there any instruction to build for aarch64 linux? Or it is more like following general build guideline and add `--build-java` flags? Or can I just used the released shared library to generate JNI so directly?

**Comment by Craigacp** [comment]
It built & the tests ran successfully on an Oracle Linux 7 box running on a 4 core Ampere Altra instance. Took a little while to get the machine setup but the build was without incident.

**Comment by lanking520** [comment]
> It built & the tests ran successfully on an Oracle Linux 7 box running on a 4 core Ampere Altra instance. Took a little while to get the machine setup but the build was without incident.

Nice! For now I will just follow the default build options to see if it can work. If build works, I can probably bundle the JNI so files together with released shared library to have some extra optimization.

**Comment by ljordan-pea** [comment]
I am trying to use onnx java runtime on an m1 as well. I found you could use x86 flavor of java and things will work. I am currently using:

amazon-corretto-15.jdk

I can't speak for the speed or whatnot, but at least my test pass now  and I can keep working :) I am not planning on deploying to macOS.

**Comment by Craigacp** [comment]
I'd be careful running ONNX Runtime under Rosetta on an M1. I'm not sure if the detection in MLAS assumes all x86 CPUs have AVX instructions or not. If it fails the detection and uses an AVX instruction these are not emulated by Rosetta and can cause an illegal instruction error which will crash the JVM if one is executed under it.

**Comment by OlegTheCat** [comment]
Hi @faxu! As of 1.12 release, I'm getting this error when using Java package on M1 Mac:

```
java.lang.UnsatisfiedLinkError: no onnxruntime in java.library.path
```

As far as I understand, M1 binaries have been placed into `ai/onnxruntime/native/osx-arm64/` directory. However corresponding Java code [is looking for](https://github.com/microsoft/onnxruntime/blob/9616ad483f5c67b36b4610a161e83e67373fb01c/java/src/main/java/ai/onnxruntime/OnnxRuntime.java#L112) `ai/onnxruntime/native/osx-aarch64/` directory. 
Is it something that has been overlooked or am I doing something wrong? Thanks!

**Comment by pwittchen** [comment]
@faxu I bumped onnxruntime in my project to `1.12.0` and I still get errors like:

```
no onnxruntime in java.library.path
java.lang.UnsatisfiedLinkError: no onnxruntime in java.library.path
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1875)
```

or

```
Could not initialize class ai.onnxruntime.OrtEnvironment
java.lang.NoClassDefFoundError: Could not initialize class ai.onnxruntime.OrtEnvironment
```

When I use my compiled `*.jar` file with onnxruntime for Mac and include it manually in the project for Mac, then it works. I guess something is wrong with the final jar or you need to release separate jars for different operating systems (Windows, Linux, Mac)

**Comment by faxu** [comment]
@leqiao-1 / @snnn can you take a look at this for Java M1 packaging?

**Comment by lanking520** [comment]
I just checked the content and found the `osx-amd64` and `osx-x64`. One of the naming of the binary is wrong. AMD64 is euqivalent to x64. @Craigacp Happy to jump in for a fix if needed.

**Comment by pwittchen** [comment]
Thanks for your comments. I created PR here: #12325 according to your suggestions. Please, review it.

**Comment by pwittchen** [comment]
My PR was closed without merging after discussion because changes need to be done in the other place. The PR no. #12335 should fix the Maven Central packaging issue. In the meantime, `*.jar` artifact compiled from the source code on the Macbook with M1 can be used during development on that machine and works fine.

**Comment by russellrc-keebo** [comment]
If I understand correctly, this should be fixed by https://github.com/microsoft/onnxruntime/pull/12335

**Comment by snnn** [comment]
We will publish a patch release to fix this.

**Comment by OlegTheCat** [comment]
Hi @snnn! Thanks for looking into this! In the newest version (1.12.1) everything works perfect! I'm able to use ONNX Runtime using Java package on M1 Mac.

**Comment by pwittchen** [comment]
@russellrc-keebo yes, it's fixed by #12335. In the version `1.12.1` this problem no longer occurs.
